feat: Expose `buttons` accessor on Tap Details events (TapDownDetails and TapUpDetails)

luanpotter opened the pull request:
Expose `buttons` accessor on Tap Details events (`TapDownDetails` and `TapUpDetails`).

Fixes https://github.com/flutter/flutter/issues/176583

Note: I chose the option of making it optional to avoid breaking anyone creating synthetic events by hand (though I doubt many people would be doing that). I am also perfectly happy with making it required - just let me know. That would be more ergonomic to the end user as the framework can always guarantee it is there but might break those edge cases of people creating these classes "by hand". A compromise would be a default value, though I think that is more misleading than it would be helping. If we are ok with the minor breaking change my preference would be to make it required. But I think adding it as optional is a much safer first step.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md

luanpotter committed:
feat: Expose `buttons` accessor on Tap Details events (TapDownDetails and TapUpDetails)

luanpotter committed:
Update packages/flutter/lib/src/widgets/text_selection.dart

Co-authored-by: Renan <6718144+renancaraujo@users.noreply.github.com>

Renzo-Olivares commented on packages/flutter/lib/src/widgets/text_selection.dart:
```diff
@@ -2432,6 +2432,7 @@ class TextSelectionGestureDetectorBuilder {
     // vs [TapGestureRecognizer.onSecondaryTapUp] instead of having to track state in
     // renderEditable. When this migration is complete we should remove this hack.
     // See https://github.com/flutter/flutter/issues/115130.
+    // TODO(luanpotter): Provide kind and buttons here once [TapDragDownDetails] has them.
```
I think if we want to expose the `*Details.buttons` API it would be preferred if this pull request handled adding them to all the `*Details` objects and not only `TapDownDetails` and `TapUpDetails`. This would be similar to how we added the `allowedButtonsFilter` API https://github.com/flutter/flutter/pull/111852 on all the recognizers in one PR. I think the benefit of doing it this way versus the alternative of multiple PRs is that the framework is not left in a inconsistent state. What do you think about this?

luanpotter committed:
feat: Add and wire in `buttons` to all event details

luanpotter commented on dev/integration_tests/flutter_gallery/lib/demo/transformations/transformations_demo_gesture_transformable.dart:
```diff
@@ -211,6 +211,8 @@ class _GestureTransformableState extends State<GestureTransformable> with Ticker
                     details.globalPosition - getOffset(context),
                     _transform,
                   ),
+                  kind: details.kind,
```
NOTE: I am also wiring kind on places that I found it missing. Please let me know if the omission was intentional in these places.

luanpotter commented on packages/flutter/lib/src/gestures/drag_details.dart:
```diff
@@ -35,6 +40,15 @@ class DragDownDetails with Diagnosticable implements PositionedGestureDetails {
+
+  /// {@macro flutter.gestures.PointerEvent.buttons}
+  ///
+  /// NOTE: this will always be set by the platform but synthetic events might
```
didn't feel great to hide a NOTE inside the macro (and it might not apply everywhere) so I left it here, but happy to pivot.

luanpotter replied:
would you still put that into a macro? it seems we are going in the direction of more customized per-event messages (such as referring to drag-specifics or referring to things with the exact name in the context they are). makes sense to just not use a macro at all?

luanpotter commented on packages/flutter/lib/src/gestures/drag_details.dart:
```diff
@@ -180,6 +209,8 @@ class DragUpdateDetails with Diagnosticable implements PositionedGestureDetails
     properties.add(DiagnosticsProperty<Duration?>('sourceTimeStamp', sourceTimeStamp));
     properties.add(DiagnosticsProperty<Offset>('delta', delta));
     properties.add(DoubleProperty('primaryDelta', primaryDelta));
+    properties.add(EnumProperty<PointerDeviceKind?>('kind', kind));
```
this was missing but I believe was a mistake

luanpotter commented on packages/flutter/lib/src/gestures/multidrag.dart:
```diff
@@ -58,6 +58,9 @@ abstract class MultiDragPointerState {
   final PointerDeviceKind kind;
 
+  /// {@macro flutter.gestures.PointerEvent.buttons}
+  final int buttons;
```
not sure if this one needs to be nullable as well - since the `kind` is not

luanpotter committed:
Fix rebase with type annotations change

luanpotter committed:
Fix lint

luanpotter committed:
Fix analysis on tap_test.dart

luanpotter committed:
Fix lint

luanpotter committed:
Add missing kind to LongPressStartDetails - all events need kind because the docs for buttons refer to it

luanpotter committed:
Add remaining missing kinds to fix buttons dartdocs issue

luanpotter commented:
@Renzo-Olivares I'm not sure the remaining failure is related, I see it is due to:
```
Retry: group retry flag allows test to run multiple times the group with retry flag the test inside it
00:29 +613 ~3 -4: /b/s/w/ir/x/w/flutter/packages/flutter_test/test/widget_tester_test.dart: hasRunningAnimations control test
00:29 +614 ~3 -4: /b/s/w/ir/x/w/flutter/packages/flutter_test/test/widget_tester_test.dart: pumping pumpFrames
00:29 +615 ~3 -4: /b/s/w/ir/x/w/flutter/packages/flutter_test/test/widget_tester_test.dart: pumping pumping
00:29 +616 ~3 -4: /b/s/w/ir/x/w/flutter/packages/flutter_test/test/widget_tester_test.dart: verifyTickersWereDisposed control test
00:29 +617 ~3 -4: /b/s/w/ir/x/w/flutter/packages/flutter_test/test/widget_tester_test.dart: can pass a View to pumpWidget when wrapWithView: false
00:29 +618 ~3 -4: /b/s/w/ir/x/w/flutter/packages/flutter_test/test/widget_tester_test.dart: wrapWithView: false does not include View
00:29 +619 ~3 -4: Some tests failed.
```
But that doesn't seem related; the tests pass locally for me. Wondering if this is some well-known or CI issue.

luanpotter commented on packages/flutter/test/gestures/debug_test.dart:
```diff
@@ -263,32 +268,45 @@ void main() {
           'globalPosition: Offset(0.0, 0.0)',
           'localPosition: Offset(0.0, 0.0)',
           'velocity: Velocity(0.0, 0.0)',
+          'kind: null',
```
intentionally left some as null to test that as well

luanpotter committed:
more fixes

dkwingsmt reviewed (commented):
I left some comments on documentation. If you agree with it, similar changes should be applied to other properties as well.

I prefer making it non-null with a default argument rather than making it nullable. Here's my reasoning:

**Categorizing the details object by origin:**
* Type 1: Platform events, constructed within framework code
* Type 2: Synthesized in tests, constructed within framework code
* Type 3: Platform events, constructed outside framework code (typically 3rd party drag recognizers)
* Type 4: Synthesized in tests, constructed outside framework code (3rd party unit tests)

**Analysis by type:**
* **Types 1 & 2**: We can handle all migrations internally, so nullability doesn't matter.
* **Type 3**: Remember that we're adding these properties because consumer classes (like `MultiTapGestureRecognizer`) need them to function correctly. Making properties nullable shifts the default value responsibility to consumer classes, which is worse because it's distributed and error-prone.
* **Type 4**: Existing tests don't rely on these properties, so any value works.

**Conclusion**: Making it nullable is strictly worse for Type 3 and equivalent for others.

**Default argument vs. required parameter:**
* **Downside of default argument**: 3rd party recognizers won't be notified to update their code with correct values, undermining the non-null benefit.
* **Downside of required parameter**: All unit tests simulating drags will break.

Since there are likely far more unit tests than 3rd party recognizers, a default argument is the best compromise. Centralizing default value logic is better than distributing it across consumer code.

luanpotter commented:
@dkwingsmt I agree with your type breakdown - very well put. My only minor concern are the Type 3 external recognizers not being aware they might need to set this if they are firing synthetic events. Distributing the default value logic might not be ideal in all cases if different use cases could otherwise have different default behaviours (intentionally). But I agree that is quite the edge case - so worth making it required, defaulted, and centralized seems like a solid plan for me.

I will update the PR accordingly when I have some spare cycles. Thank you so much for initial the review!

luanpotter committed:
Address comments; make kind and button mandatory with default values; enhance comments; remove comment macro

luanpotter committed:
Fix analyzer

luanpotter committed:
Fix consequences of making kind not-null

luanpotter committed:
Fix consequences of making kind not-null

luanpotter committed:
Fix inconsistent test

luanpotter commented on packages/flutter_test/test/controller_test.dart:
```diff
@@ -239,7 +239,7 @@ void main() {
         expect(logs[i], 'move $b');
       } else {
-        expect(logs[i], 'up 0');
+        expect(logs[i], 'up $b');
```
i'm not sure why this test was expecting 0 here instead of the actual button (2). I think it was just testing the wrong previous behaviour incorrectly

luanpotter committed:
Fix inconsistent tests

luanpotter committed:
Fix tests

luanpotter commented on packages/flutter_test/test/controller_test.dart:
```diff
@@ -403,6 +403,9 @@ void main() {
+        // The PointerData received from the engine with
+        // change = PointerChange.up, which translates to PointerUpEvent,
+        // doesn't provide the button field.
         expect(logs[i], 'up 0');
```
this is from https://github.com/flutter/flutter/blob/7889cccfbe970b5470c3ea56f9fddc2d77e23f4b/packages/flutter_test/lib/src/controller.dart#L1800-L1803

seems pretty intentional; if the line were uncommented there, this could follow the other tests

luanpotter commented on packages/flutter/lib/src/widgets/selectable_region.dart:
```diff
@@ -763,14 +763,14 @@ class SelectableRegionState extends State<SelectableRegion>
         switch (defaultTargetPlatform) {
           case TargetPlatform.iOS:
-            if (kIsWeb && details.kind != null && !_isPrecisePointerDevice(details.kind!)) {
+            if (kIsWeb && !_isPrecisePointerDevice(details.kind)) {
```
more cascading consequences of `kind` being non-null

luanpotter committed:
Fix format

luanpotter committed:
Fix formatting on more files

luanpotter commented:
@dkwingsmt happy new year! I've updated the PR according to the comments, please lmk what you think.
* made kind and buttons non-nullable with default values, making necessary adjustments
* updated docs, removed the macro, tailored the wording for each different event type

Renzo-Olivares reviewed (commented):
@luanpotter thank you for your patience and contribution! Overall I agree with the approach of making `buttons` non-nullable with a default value. However, i'm not so sure about changing the existing nullable `kind` to non-nullable. This may be a breaking change, if a developer had some fallback logic that runs when kind is null, this logic will no longer run since it now has a default value. I think if we want to explore making `kind` `non-null` we should do this in a separate PR.

dkwingsmt reviewed (commented):
Thank you for the work. I think it's on the right direction!

luanpotter commented on packages/flutter/lib/src/widgets/scrollbar.dart:
```diff
@@ -1735,6 +1735,8 @@ class RawScrollbarState<T extends RawScrollbar> extends State<T> with TickerProv
     final details = DragStartDetails(
```
I might also extract this to a PR we can merge first

luanpotter committed:
Address more comments; keep existing kinds nullable and new kinds/buttons with default values

luanpotter committed:
Address remaining comments

luanpotter committed:
Fix rebase on text_selection.dart

luanpotter committed:
Fix rebase on packages/flutter/lib/src/widgets/selectable_region.dart

luanpotter committed:
Fix lint

luanpotter commented:
@dkwingsmt @Renzo-Olivares I think I addressed all comments but I realize this PR is getting a bit out of hand in terms of size; in particular [this change](https://github.com/flutter/flutter/pull/176968#discussion_r2695986075) drove me over the fence and made me prefer to split it up. I am keeping this PR with all the changes for now so you can analyze as a whole package (I will rebase later), but I think it would be safer to merge some pieces first (since I was biting the bullet and extracting the scrollbar change, I went ahead and extracted a couple more). The bulk of the changes will still be this PR as they are in a way inseparable unless we want to land some form of "in-between" state, which I would rather not.

But these changes are very tangential, and I think make sense separately:

* https://github.com/flutter/flutter/pull/181110 - addressing the comment about the scrollbars
* https://github.com/flutter/flutter/pull/181109 - pass-by improvement on error messages
* https://github.com/flutter/flutter/pull/181113 - addresses the data class comment

I also started to separate a change to make all `EnumProperty` of `Foo` and not `Foo?`, though I think this is much more of an open-ended philosophical question and totally not necessary to merge this PR; so I am listing it separately but entirely up to you guys what you prefer there.

* https://github.com/flutter/flutter/pull/181112

Hopefully this will make it easier as you can review these 3 (or 4) much easier PRs and then I can rebase this one and make it at least a bit smaller (will still be definitely 10+ files though). I think we are honing in the solution thanks to your comments and guidance so I think I have much better picture of everything in my head now.

Thanks again for all the reviews and sorry for the back and forth :pray:

dkwingsmt commented:
I'm converting this to a draft since it is blocked by a few PRs. We'll resume this when all blockers are landed.
